src/auth/bearer/UserRequest.cc

```diff
@@ -0,0 +1,264 @@
+        usr->username(userNote);
+        auth_user_request->denyMessage("Login successful");
+        const int64_t ttl = (ttlNote ? strtoll(ttlNote, nullptr, 10) : -1);
+        usr->token->expires = current_time.tv_sec + ttl;
+        usr->expiretime = max(usr->expiretime, usr->token->expires);
+        usr->credentials(Auth::Ok);
+        usr->addToNameCache();
+        debugs(29, 4, "Successfully validated user via Bearer. Username " << auth_user_request->user()->username());
+    }
+    break;
+
+    case Helper::Error: {
+        const char *messageNote = reply.notes.findFirst("message");
+        const char *ttlNote = reply.notes.findFirst("ttl");
+
+        /* authentication failure (wrong password, etc.) */
+        if (messageNote)
+            auth_user_request->denyMessage(messageNote);
+        else
+            auth_user_request->denyMessage("Bearer Authentication denied with no reason given");
+
+        Auth::Bearer::User *usr = dynamic_cast<Auth::Bearer::User *>(auth_user_request->user().getRaw());
+
+        const int64_t ttl = (ttlNote ? strtoll(ttlNote, nullptr, 10) : -1);
+        usr->token->expires = current_time.tv_sec + ttl;
+        usr->expiretime = max(usr->expiretime, usr->token->expires);
+        usr->credentials(Auth::Failed);
+        debugs(29, 4, "Failed validating user via Bearer. Result: " << reply);
+    }
+    break;
+
+    case Helper::Unknown:
+        debugs(29, DBG_IMPORTANT, "ERROR: Bearer Authentication Helper crashed (" << reply.reservationId << ")");
+        /* continue to the next case */
+
+
+    case Helper::TimedOut:
+        /* continue to the next case */
+
+    case Helper::TT:
+        /* continue to the next case */
+
+    case Helper::BrokenHelper: {
+        const char *errNote = reply.notes.findFirst("message");
+        if (reply.result == Helper::Unknown)
+            auth_user_request->denyMessage("Internal Error");
+        else if (errNote)
+            auth_user_request->denyMessage(errNote);
+        else
+            auth_user_request->denyMessage("Bearer Authentication failed with no reason given");
+        auth_user_request->user()->credentials(Auth::Failed);
+        debugs(29, DBG_IMPORTANT, "ERROR: Bearer Authentication validating user. Result: " << reply);
+    } // break;
+    }
+
+    // Notify all waiting transactions of the result
+    void *cbdata;
+    if (cbdataReferenceValidDone(r->data, &cbdata))
+        r->handler(cbdata);
+
+    cbdataReferenceDone(r->data);
+
+    Auth::Bearer::User *local_usr = dynamic_cast<Auth::Bearer::User *>(auth_user_request->user().getRaw());
+    while (local_usr->queue) {
+        if (cbdataReferenceValidDone(local_usr->queue->data, &cbdata))
+            local_usr->queue->handler(cbdata);
+
+        Auth::QueueNode *tmpnode = local_usr->queue->next;
+        local_usr->queue->next = nullptr;
+        delete local_usr->queue;
+
+        local_usr->queue = tmpnode;
+    }
+
+    delete r;
```

Comment:
That "nearly" is critical difference of what type of object is being accessed. Despite performing the same logical operations the object and handler types are different. Refactoring the queue to avoid this is out of scope here.